Remove local userpaths from error reporting

derekblank opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

Resolves Automattic/dotcom-forge#6701

## Proposed Changes
Prevents local userpaths from being logged to error reporting.

- Adds a new function `sanitizeUserpath` to sanitize-for-logging utilities
- Updates user-data `console.log` events to remove matching userpaths from logged events

Currently, the userpath pattern matcher matches `/Users/username/Library/...`, which is the expected userpath on macOS. When release support for other operating systems is added, this userpath matcher can be expanded to cover the new OS userpath patterns.  

## Testing Instructions

To test, observe that usernames are not included in userpaths when **logging** user-data functions:
- [migrateUserData](https://github.com/Automattic/studio/blob/7996f15d3bf281e6efcf4f538b5bd6d11c010550/src/storage/user-data.ts#L11)
- [loadUserData](https://github.com/Automattic/studio/blob/7996f15d3bf281e6efcf4f538b5bd6d11c010550/src/storage/user-data.ts#L31)
- [saveUserData](https://github.com/Automattic/studio/blob/7996f15d3bf281e6efcf4f538b5bd6d11c010550/src/storage/user-data.ts#L75)


### Sentry
These functions can be observed on any Sentry session sending these events:

<img width="765" alt="Screenshot 2024-05-02 at 10 30 34 pm" src="https://github.com/Automattic/studio/assets/643285/b2899ba9-8bdb-4f6a-87c5-0164a5bf081c">

### Local Development
Another more straightforward way to test the changes is to observe the same events when logged while running Studio locally:
<img width="547" alt="Screenshot 2024-05-02 at 10 31 28 pm" src="https://github.com/Automattic/studio/assets/643285/3cdb72bf-339d-44e8-a3db-ffae3dce7e49">





## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Have you checked for TypeScript, React or other console errors?

derekblank committed:
Add sanitizeUserpath util

derekblank committed:
Add sanitizeUserpath tests to sanitize-for-logging

derekblank committed:
Update user-data to use sanitizeUserpath

derekblank committed:
Update migrateUserData to use sanitizeUserpath

derekblank committed:
Update loadUserData to use sanitizeUserpath on logging errors

derekblank commented on src/lib/sanitize-for-logging.ts:
```diff
@@ -28,3 +28,14 @@ export function sanitizeUnstructuredData( data: string ): string {
+	// Matches expected user path on macOS
+	if ( typeof path === 'string' && /\/Users\/[^/]+\/Library/.test( path ) ) {
+		return path.replace( /\/Users\/[^/]+(\/Library)/, '~$1' );
+	}
```
Currently, the userpath pattern matcher matches `/Users/{username}/../..`, which is the expected userpath on macOS (presently, the only OS that has been released).

When support for other operating systems is added, this userpath matcher can be expanded to cover the new OS userpath patterns.  Also noting that this util will return the original string if not matching the pattern.

Open to feedback if we want to include other OS userpath matchers now, however (Windows, Linux).

derekblank committed:
Update userpath pattern matcher to accept a wider filepath match

dcalhoun commented on src/lib/sanitize-for-logging.ts:
```diff
@@ -28,3 +28,14 @@ export function sanitizeUnstructuredData( data: string ): string {
+export function sanitizeUserpath( path: string ): string {
+	const userPathRegex = /^\/Users\/([^/]+)\//;
+	if ( typeof path === 'string' && userPathRegex.test( path ) ) {
+		return path.replace( userPathRegex, '~/' );
```
I think replacing with a tilde is sensible. 

I wondered if we should instead rely upon the existing practice of inserting `REDACTED` instead, e.g. `/Users/REDACTED/`. That might more clearly communicate an action was taken on the string. 

Also, when we support other OSs, I supposed need to update it to use the equivalent for macOS' tilde usage — e.g., my understanding is that Windows might use `C:/`. 

Not necessarily requesting a change. Merely sharing thoughts.

dcalhoun commented on src/lib/sanitize-for-logging.ts:
```diff
@@ -28,3 +28,14 @@ export function sanitizeUnstructuredData( data: string ): string {
+// Returns the original path if it does not match the pattern or if it is not a string.
+export function sanitizeUserpath( path: string ): string {
+	const userPathRegex = /^\/Users\/([^/]+)\//;
+	if ( typeof path === 'string' && userPathRegex.test( path ) ) {
```
Using both `test` and `replace` may be unnecessary. If `replace` doesn't find a pattern match, it would have no impact. Are we using `test` to guard against something in particular? 

```suggestion
	if ( typeof path === 'string' ) {
```

dcalhoun replied:
Also, I suppose our usage of TypeScript should, in theory, negate a need to check with `typeof`, correct?

derekblank replied:
Initially, I had this regex matcher as part of `sanitizeForLogging` function itself, but decided it was too convoluted to accept both key/value objects and strings. Using a dedicated function for `sanitizeUserpath` is more readable, and these extra typechecks are no longer needed. Updated below, and thanks for catching.

dcalhoun reviewed (approved):
The test plan succeeded for me. Thank you for address this. 🙇🏻 

I note that logged server details include usernames in the `projectPath` and `wpContentPath` attributes. However, I was unable to find an instance of those values in Sentry. So, we may not need to scrub those. I'm not sure. WDYT? 

https://github.com/Automattic/studio/blob/c545108229b03b0081359f4b727d8af49cd76181/src/site-server.ts#L86

derekblank committed:
Merge trunk and resolve conflicts

derekblank committed:
Update userpath matcher to match Windows and Unix strings

derekblank committed:
Update sanitizeUserpath tests

derekblank committed:
Remove unnecessary typecheck

derekblank commented:
> I wondered if we should instead rely upon the existing practice of inserting REDACTED instead, e.g. /Users/REDACTED/. That might more clearly communicate an action was taken on the string.
> 
> Also, when we support other OSs, I supposed need to update it to use the equivalent for macOS' tilde usage — e.g., my understanding is that Windows might use C:/.

@dcalhoun Thanks for the feedback - I think these are good suggestions. 

Given that Windows support is under active development and release is intended to be pending soon-ish, I think it's a good idea to go ahead and accommodate Windows and Unix together in this PR.

The [test cases](https://github.com/Automattic/studio/pull/78/files#diff-806035b0b3c62f79b3e07fa9f15f7adf2a94ec745e65ec76a0d0cecd642de06aR85-R104) probably make the intent clearer, but I've made the following changes:

- Updated the regex pattern matcher to account for forward/backslash differences Windows and macOS userpaths, as well as the Windows drive letter.
- Used `[REDACTED]` in place of `~` to match the spirit of our other sanitize utils.
- Updated `sanitizeUserpath` to not sanitize the userpath in development mode -- It's redacted when sent to Sentry, but perhaps users may want to see their unredacted userpath when developing locally? (Open to feedback.)
- Removed unneeded typechecks.

Let me know if you have further thoughts on this approach.

derekblank commented:
> I note that logged server details include usernames in the `projectPath` and `wpContentPath` attributes. However, I was unable to find an instance of those values in Sentry. So, we may not need to scrub those. I'm not sure. WDYT?

I looked at these as well, and my understanding is that the filtering here is handled either by [sanitizeForLogging](https://github.com/Automattic/studio/blob/abfe4dabdaaa4359d6b9fa7f0428db9caae412dc/src/lib/sanitize-for-logging.ts#L3-L22), or by [sanitizeUnstructuredData](https://github.com/Automattic/studio/blob/abfe4dabdaaa4359d6b9fa7f0428db9caae412dc/src/lib/sanitize-for-logging.ts#L26-L29). 

Initially, I attempted to leverage either of these functions to also handle the regex matching for the userpath, but ultimately decided that `sanitizeUserpath` was best served with its own function as it did not utilize a key/value pair, unlike the other two sanitization functions. (This is also where the extra typechecks in the initial function came from -- thanks for [calling out](https://github.com/Automattic/studio/pull/78#discussion_r1587649556) that those were no longer needed.)

katinthehatsite reviewed (approved):
The testing steps succeeded for me. Here is what I did:

* First, I tested on trunk to confirm the issue - the userpath was not redacted there:

<img width="697" alt="Screenshot 2024-05-03 at 10 14 53 AM" src="https://github.com/Automattic/studio/assets/25575134/7a4e3d93-a1ab-408a-adde-f7c558e6e758">

* Next, I switched to the branch and observed that the userpath is not redacted either since I am running the path in the development mode. 

* I removed the following condition:

```
if ( process.env.NODE_ENV !== 'development' ) {
```
And then ran `nvm use && npm install && npm start` and the userpath was redacted as expected:

<img width="689" alt="Screenshot 2024-05-03 at 10 19 20 AM" src="https://github.com/Automattic/studio/assets/25575134/0fdd3788-a539-4570-a471-2a7d2d18bf11">

The changes seem to work on my end on macOs. Please note that I have not tested this on Windows so I am only confirming the functionality for macOS.

katinthehatsite commented:
> Updated sanitizeUserpath to not sanitize the userpath in development mode -- It's redacted when sent to Sentry, but perhaps users may want to see their unredacted userpath when developing locally?

I don't have a very strong preference for this one. Actually, I like seeing the path when the app is in the dev mode in case I need to locate something quickly e.g. `appdata-v1.json` (especially when testing many PRs and needing to change many things). 👍
